dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/programrule/implementers/AssignValueImplementer.java

```diff
@@ -159,10 +160,15 @@ private boolean isTheSameValue( EventActionRule actionRule, TrackerPreheat prehe
 
     private boolean isTheSameValue( EnrollmentActionRule actionRule, TrackerPreheat preheat )
     {
-        TrackedEntityAttribute attribute = preheat.get( TrackedEntityAttribute.class, actionRule.getField() );
+        TrackedEntityAttribute attribute = preheat.getTrackedEntityAttribute( actionRule.getField() );
         String value = actionRule.getValue();
+        // NOTE: since rule engine has no notion of idSchemes the attribute
+        // identifiers have to be mapped to UIDs
+        // here we "escape" the safety of MetadataIdentifier and assume it
+        // contains a UID; thus directly comparing
+        // the identifier String to the field String
         Optional<Attribute> optionalAttribute = actionRule.getAttributes().stream()
-            .filter( at -> at.getAttribute().equals( actionRule.getField() ) )
+            .filter( at -> at.getAttribute().getIdentifier().equals( actionRule.getField() ) )
```

Comment:
EnrollmentActionRule.field is a String. So this string based identifier comparison only works for different idSchemes because I mapped the payload attributes to UIDs in `mergeAttributes` above.